show.md

```diff
@@ -0,0 +1,207 @@
+# Showing and hiding geometry (construction geometry, assemblies)
+
+I propose an 'implicit show' syntax (with an explicit version as an alternative), where geometry which is consumed (by assigning into a variable, passing to a function, etc.) is treated as construction geometry and geometry which escapes to the top-level of a module is rendered (or becomes part of an assembly).
+
+## Background and motivation
+
+Most CAD software has the concept of *construction geometry* which is geometry used for constructing other geometry but which is not itself rendered. We currently don't have explicit support for this in KittyCAD, but the engine does make some choices about which geometry is rendered. In KCL, non-rendered geometry might take other forms too, e.g., geometry in functions or in imported modules. It's also likely to be used in many places, e.g., creating a single instance of an object and then using a pattern to create multiple instances, where we don't want to render the original instance.
+
+One particular question to answer is how geometry should work with imported modules (which I envisage to be the way that KCL supports assemblies). To recap the current and [proposed](https://github.com/KittyCAD/kcl-experiments/pull/12) syntax:
+
+- modules are declared in a separate file
+- exported functions of a file can be imported using `import foo from 'bar.kcl'`
+- the module itself (i.e., the geometry defined at the top-level of the file) can be imported using `import 'bar.kcl'` (creates a single object bound to the variable `bar`, which should be able to be used as an assembly)
+
+Currently, objects are effectively rendered by virtue of the side effects of execution (issuing API calls to the engine). Therefore, exactly how all the above might work is a bit vague. I'll spell some of it out here, focussing on the user-facing syntax. More of the fundamentals are discussed in the [foundations design doc](foundations.md), including side-effects.
+
+
+## Proposal: implicit `show`
+
+The basics of the proposal is that an object that is defined at the top level is rendered by the engine unless it is passed into a function, used in a pipeline, or assigned into a variable (or used in some other way we add to KCL in the future). Using the result of a function or a variable at the top level will render it. E.g.,
+
+```
+// Result of function call, rendered
+makeSphere()
+
+// Result of pipeline, rendered (nothing intermediate in the pipeline is rendered)
+startSketchOn(...)
+  |> startPath()
+  |> ...
+  |> close()
+  |> extrude()
+
+// As above, but assigned into a variable, nothing is rendered
+object = startSketchOn(...)
+  |> startPath()
+  |> ...
+  |> close()
+  |> extrude()
+
+// Variable is used, result of pipeline is rendered
+object
```

Comment:
Just noting that even if we have an explicit `show` keyword or function, we'd have the same issue. We should track what is shown I guess.

Reply:
This is a good question. I think this is up to the engine rather than KCL really (although of course there is the question of what we send the engine). I'm also not sure if there is a difference between displaying twice or once? If it's not named or tagged, then it can't be referred to and if it's at exactly the same spot, the image shouldn't change. So I would argue that how many actual draw calls are made is just an optimisation that cannot be observed. In other words, KCL should not track what has been displayed